src/Shell.cpp

```diff
@@ -0,0 +1,206 @@
+#include <thread>
+#include <readline/readline.h>
+#include <readline/history.h>
+
+using namespace std;
+namespace fs = filesystem;
+
+
+
+// this function takes a string and a vector of strings as input and parses
+// the string into tokens and stores them in the vector
+void Shell::parseCommand(const string &input, vector<string> &args) 
+{
+    stringstream ss(input);
+    string token;
+    while (ss >> token) {
+        args.push_back(token);
+    }
+}
+
+void Shell::addJob(pid_t pid, const string& cmd) 
+{
+    Job job = {pid, cmd, "Running"};
+    bgJobs[pid] = job;
+}
+
+// this function goes over the bgJobs map and updates the status of each job
+void Shell::updateJobStatus() 
+{
+    for (auto it = bgJobs.begin(); it != bgJobs.end();) {
+        int status;
+        pid_t result = waitpid(it->first, &status, WNOHANG);
+        if (result == 0) {
+            it->second.status = "Running";
+            ++it;
+        } else if (result == -1) {
+            perror("waitpid failed");
+            it = bgJobs.erase(it);
+        } else {
+            if (WIFEXITED(status) || WIFSIGNALED(status)) // this signal is sent to a process when its controlling terminal is closed
+                it = bgJobs.erase(it); // no use in a stopped job in a currently running job list
+            else 
+                ++it;
+            
+        }
+    }
+}
+
+// this simply goes over the pid - string map and prints out the pid, command and status of each job
+void Shell::myJobs() 
+{
+    updateJobStatus();
+    for (const auto& job : bgJobs) {
+        cout << "PID: " << job.second.pid << ", Command: " << job.second.command << ", Status: " << job.second.status << endl;
+    }
+}
+
+
+// my main method for this shell program
+// it goes over the input string and processes it accordingly
+// it checks for in program commands and also path commands
+void Shell::inputLoop() 
+{
+    while (true) 
+    {
+        try
+        {
+            char* input = readline("");
+            if (input == nullptr) break;
+            
+
+
+            string command(input);
+            add_history(command.c_str());
+            Utils::addCommandToHistory(command);
+        
+
+
+            if (command == "") continue;
+            if (command == "exit") break;
+            if (command == "myjobs") 
+            {
+                myJobs();
+                continue;
+            }
+            if (command == "myhistory") 
+            {
+                myHistory();
+                continue;
+            }
+            if (command == "help") 
+            {
+                help();
+                continue;
+            }
+            if (command == "who") 
+            {
+                who();
+                continue;
+            }
+
+            command = Utils::parseEnvironmentVariables(command);
+
+            //process command
+            vector<string> args;
+            parseCommand(command, args);
+            bool runInBackground = false;
```

Comment:
Choose descriptive names for your boolean variables, with prefixes like `is`, `can`, `should`, etc.
(for example: `shouldSave`, `isValid`, `canProcess`, etc.)